src/cursor/cur_backup.c

```diff
@@ -216,48 +239,67 @@ __backup_start(
-	 * could start a hot backup that would race with an already-started
-	 * checkpoint.
-	 *
-	 * We are holding the checkpoint and schema locks so schema operations
-	 * will not see the backup file list until it is complete and valid.
-	 */
-	__wt_writelock(session, &conn->hot_backup_lock);
-	conn->hot_backup = true;
-	conn->hot_backup_list = NULL;
-	__wt_writeunlock(session, &conn->hot_backup_lock);
+	if (other == NULL) {
+		/*
+		 * The hot backup copy is done outside of WiredTiger, which
+		 * means file blocks can't be freed and re-allocated until the
+		 * backup completes. The checkpoint code checks the backup flag,
+		 * and if a backup cursor is open checkpoints aren't discarded.
+		 * We release the lock as soon as we've set the flag, we don't
+		 * want to block checkpoints, we just want to make sure no
+		 * checkpoints are deleted.  The checkpoint code holds the lock
+		 * until it's finished the checkpoint, otherwise we could start
+		 * a hot backup that would race with an already-started
+		 * checkpoint.
+		 *
+		 * We are holding the checkpoint and schema locks so schema
+		 * operations will not see the backup file list until it is
+		 * complete and valid.
+		 */
+		__wt_writelock(session, &conn->hot_backup_lock);
+		conn->hot_backup = true;
+		conn->hot_backup_list = NULL;
+		__wt_writeunlock(session, &conn->hot_backup_lock);
 
-	/* We're the lock holder, we own cleanup. */
-	F_SET(cb, WT_CURBACKUP_LOCKER);
+		/* We're the lock holder, we own cleanup. */
+		F_SET(cb, WT_CURBACKUP_LOCKER);
 
-	/*
-	 * Create a temporary backup file.  This must be opened before
-	 * generating the list of targets in backup_uri.  This file will
-	 * later be renamed to the correct name depending on whether or not
-	 * we're doing an incremental backup.  We need a temp file so that if
-	 * we fail or crash while filling it, the existence of a partial file
-	 * doesn't confuse restarting in the source database.
-	 */
-	WT_ERR(__wt_fopen(session, WT_BACKUP_TMP,
-	    WT_FS_OPEN_CREATE, WT_STREAM_WRITE, &cb->bfs));
+		/*
+		 * Create a temporary backup file.  This must be opened before
+		 * generating the list of targets in backup_uri.  This file will
+		 * later be renamed to the correct name depending on whether or
+		 * not we're doing an incremental backup.  We need a temp file
+		 * so that if we fail or crash while filling it, the existence
+		 * of a partial file doesn't confuse restarting in the source
+		 * database.
+		 */
+		WT_ERR(__wt_fopen(session, WT_BACKUP_TMP,
+		    WT_FS_OPEN_CREATE, WT_STREAM_WRITE, &cb->bfs));
+	} else
+		WT_ASSERT(session, WT_SESSION_BACKUP_CURSOR);
 	/*
 	 * If targets were specified, add them to the list. Otherwise it is a
 	 * full backup, add all database objects and log files to the list.
 	 */
 	target_list = false;
 	WT_ERR(__backup_uri(session, cfg, &target_list, &log_only));
+	/*
+	 * We're either a top-level cursor, or if a duplicate cursor the target
+	 * must be log only.
+	 */
```

Comment:
What's to prevent user from making a dup cursor that is not configured log_only?  That is, shouldn't the assertion be a check instead?

Reply:
Fixed.